lib/hope-tab-list.js

```diff
@@ -0,0 +1,730 @@
+    }, current, 'transitionend');
+  },
+
+  onAddClick() {
+    this.open({
+      title: 'Home',
+      url: 'homescreen.gaiamobile.org'
+    });
+  },
+
+  getSnapHeight() {
+    return window.innerHeight -
+      PREVIEW_HEIGHT -
+      GUTTER_HEIGHT -
+      BAR_HEIGHT;
+  },
+
+  getTabHeight() {
+    return Math.max(
+      TAB_HEIGHT,
+      this.getSnapHeight() / this.els.tabs.length
+    );
+  },
+
+  getScrollTop() {
+    return this.els.inner.scrollTop;
+  },
+
+  scrollTo(top, options={}) {
+    return new Promise(resolve => {
+      this.els.inner.scrollTo({
+        top: top,
+        behavior: options.instant ? 'auto' : 'smooth'
+      });
+
+      setTimeout(resolve, options.wait || 300);
+    });
+  },
+
+  setupSnapping() {
+    var snapHeight = this.getSnapHeight();
+    this.els.inner.style.scrollSnapPointsY = `repeat(${snapHeight}px)`;
+  },
+
+  layout() {
+    return new Promise(resolve => {
+      debug('layout');
+
+      var current = this.els.tabs[0];
+      var content = current.querySelector('hope-tab');
+      var snapHeight = this.getSnapHeight();
+      var container = this.els.container;
+
+      this.cleanUp(current);
+
+      content.active = !this.inTabsView;
+
+      current.classList.add('current');
+      current.style.height = `calc(100vh + ${snapHeight}px)`;
+      current.style.zIndex = 1;
+      current.style.top = 0;
+
+      var tabs = this.els.tabs.slice(1);
+      var tabHeight = this.getTabHeight();
+      var top = window.innerHeight - GUTTER_HEIGHT;
+
+      for (var i = 0; i < tabs.length; i++) {
+        var tab = tabs[i];
+
+        tab.style.height = tabHeight + 'px';
+        tab.style.top = top + 'px';
+        tab.style.zIndex = i + 2;
+
+        top += tabHeight - 4;
+        this.cleanUp(tab);
+      }
+
+      var newHeight = Math.max(snapHeight * 2, top);
+      var scrollTop = this.getScrollTop();
+
+      if ((scrollTop + window.innerHeight) < newHeight) {
+        updateContainer();
+        return;
+      }
+
+      this.scrollTo(scrollTop - TAB_HEIGHT)
+        .then(updateContainer);
+
+      function updateContainer() {
+        container.style.height = newHeight + 'px';
+        resolve();
+      }
+    });
+  },
+
+  open(options) {
+    if (this.opening) return;
+    this.opening = true;
+    debug('open');
+
+    var snapHeight = this.getSnapHeight();
+    var tab = this.create(options);
+    var previous = this.els.tabs[0];
+    var content = previous.firstElementChild;
+
+    tab.style.top = snapHeight - GUTTER_HEIGHT + 'px';
+    tab.style.height = window.innerHeight + 'px';
+    tab.style.zIndex = 0;
+
+    this.showFirstTab()
+      .then(() => {
+        return scheduler.mutation(() => {
+          debug('attaching new tab');
+          this.insertBefore(tab, previous);
+          this.els.tabs.unshift(tab);
+
+          content.active = false;
+          previous.classList.remove('current');
+          previous.style.top = snapHeight + 'px';
```

Comment:
When adding a timeout in the chain after this step I realized there's a small jump.
I think we need to subtract gutter height here.